risc0/zkvm/r0vm/src/main.rs

```diff
@@ -0,0 +1,138 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use clap::Parser;
+use std::fs;
+use std::io::Write;
+
+use risc0_zkvm_host::{MethodId, Prover, Receipt, DEFAULT_METHOD_ID_LIMIT};
+
+/// Generates a MethodID for a given RISC-V ELF binary.
+#[derive(Parser)]
+#[clap(about, version, author)]
+struct Args {
+    /// The ELF file to run
+    #[clap(long)]
+    elf: String,
+
+    /// MethodID file; created if it doesn't exist.
+    #[clap(long)]
+    method_id: Option<String>,
+
+    /// Proof output file.
+    #[clap(long)]
+    proof: Option<String>,
+
+    /// File to read initial input from.
+    #[clap(long)]
+    initial_input: Option<String>,
+
+    /// Display verbose output.
+    #[clap(short, long, action = clap::ArgAction::Count)]
+    verbose: u8,
+
+    /// Limit the number of hash table entries to compute.
+    #[clap(short, long, default_value_t = DEFAULT_METHOD_ID_LIMIT)]
+    limit: u32,
+}
+
+fn read_method_id(
+    verbose: u8,
+    elf_file: &str,
+    method_id_file: &Option<String>,
+) -> Option<MethodId> {
+    let elf_mtime = fs::metadata(elf_file).ok()?.modified().ok()?;
+    let id_mtime = fs::metadata(method_id_file.as_ref()?)
+        .ok()?
+        .modified()
+        .ok()?;
+
+    if elf_mtime > id_mtime {
+        return None;
+    }
+
+    let id = MethodId::from_slice(&fs::read(method_id_file.as_ref()?).ok()?).ok()?;
+
+    // TODO(nils): Check to make sure the limit is the same as the one
+    // that was saved.
+
+    if verbose > 0 {
+        println!(
+            "Successfully read method id from {}",
+            method_id_file.as_ref().unwrap()
+        );
+    }
+
+    Some(id)
+}
+
+fn main() {
+    let args = Args::parse();
+    let elf_contents = fs::read(&args.elf).unwrap();
+
+    if args.verbose > 0 {
+        eprintln!(
+            "Read {} bytes of ELF from {}",
+            elf_contents.len(),
+            &args.elf
+        );
+    }
+
+    let method_id: MethodId = read_method_id(args.verbose, &args.elf, &args.method_id)
+        .unwrap_or_else(|| {
+            if args.verbose > 0 {
+                eprintln!("Computing method id");
+            }
+            let computed = MethodId::compute(&elf_contents, args.limit).unwrap();
+            if let Some(method_id_file) = args.method_id {
+                std::fs::write(&method_id_file, computed.as_slice().unwrap()).unwrap();
+                if args.verbose > 0 {
+                    eprintln!("Saved method id to {}", method_id_file);
+                }
+            }
+            computed
+        });
```

Comment:
I'm not sure how I feel about computing the MethodID on the fly. This seems like a security issue?

@jbruestle

Reply:
We can always drop this in the future if it's an issue.